File: lib/dala/gpu/command.ex
defmodule Dala.Gpu.Command do
  @moduledoc """
  Encodes render commands to binary format for the Rust command queue.

  Binary format:
  - 1 byte: command type
  - N bytes: command-specific data

  Command types:
  - 0x01: Clear (4 bytes RGBA)
  - 0x02: FillRect (16 bytes: x,y,w,h as u32 LE + 4 bytes RGBA)
  - 0x03: DrawLine (16 bytes: x1,y1,x2,y2 as i32 LE + 4 bytes RGBA)
  - 0x04: Blit (12 bytes: sprite_id as u64 LE + x,y as u32 LE)
  - 0x05: Present (0 bytes)
  - 0x06: Resize (8 bytes: width, height as u32 LE)
  - 0x07: LoadSprite (16 bytes: id as u64 LE + w,h as u32 LE + pixel data)
  - 0x08: RemoveSprite (8 bytes: id as u64 LE)
  """

  @type color :: atom() | {0..255, 0..255, 0..255} | {0..255, 0..255, 0..255, 0..255}

  # ── Command encoding ──────────────────────────────────────────────────────

  @doc "Encode a clear command that fills the entire surface with a solid color."
  @spec encode_clear(color()) :: binary()
  def encode_clear(color) do
    rgba = color_to_rgba(color)
    <<0x01, rgba::binary>>
  end

  @doc "Encode a fill_rect command that fills a rectangle with a solid color."
  @spec encode_fill_rect(
          non_neg_integer(),
          non_neg_integer(),
          non_neg_integer(),
          non_neg_integer(),
          color()
        ) :: binary()
  def encode_fill_rect(x, y, w, h, color) do
    rgba = color_to_rgba(color)

    <<0x02, x::unsigned-little-32, y::unsigned-little-32, w::unsigned-little-32,
      h::unsigned-little-32, rgba::binary>>
  end

  @doc "Encode a draw_line command that draws a line between two points."
  @spec encode_draw_line(integer(), integer(), integer(), integer(), color()) :: binary()
  def encode_draw_line(x1, y1, x2, y2, color) do
    rgba = color_to_rgba(color)

    <<0x03, x1::signed-little-32, y1::signed-little-32, x2::signed-little-32,
      y2::signed-little-32, rgba::binary>>
  end

  @doc "Encode a blit command that draws a loaded sprite at the given position."
  @spec encode_blit(non_neg_integer(), integer(), integer()) :: binary()
  def encode_blit(sprite_id, x, y) do
    <<0x04, sprite_id::unsigned-little-64, x::signed-little-32, y::signed-little-32>>
  end

  @doc "Encode a present command that flushes the command queue and updates the GPU texture."
  @spec encode_present() :: binary()
  def encode_present do
    <<0x05>>
  end

  @doc "Encode a resize command that changes the surface dimensions."
  @spec encode_resize(non_neg_integer(), non_neg_integer()) :: binary()
  def encode_resize(width, height) do
    <<0x06, width::unsigned-little-32, height::unsigned-little-32>>
  end

  @doc "Encode a load_sprite command that uploads pixel data into the texture atlas."
  @spec encode_load_sprite(non_neg_integer(), binary(), non_neg_integer(), non_neg_integer()) ::
          binary()
  def encode_load_sprite(id, rgba_data, width, height) do
    <<0x07, id::unsigned-little-64, width::unsigned-little-32, height::unsigned-little-32,
      rgba_data::binary>>
  end

  @doc "Encode a remove_sprite command that frees a sprite from the texture atlas."
  @spec encode_remove_sprite(non_neg_integer()) :: binary()
  def encode_remove_sprite(id) do
    <<0x08, id::unsigned-little-64>>
  end

  @doc "Encode a load_image command that uploads an image as a GPU texture."
  @spec encode_load_image(non_neg_integer(), binary(), non_neg_integer(), non_neg_integer()) :: binary()
  def encode_load_image(id, rgba_data, width, height) do
    <<0x17, id::unsigned-little-64, width::unsigned-little-32, height::unsigned-little-32,
      rgba_data::binary>>
  end

  @doc "Encode a remove_image command that frees an image texture."
  @spec encode_remove_image(non_neg_integer()) :: binary()
  def encode_remove_image(id) do
    <<0x18, id::unsigned-little-64>>
  end

  @doc "Encode a dispatch_compute command that runs a GPU compute shader."
  @spec encode_dispatch_compute(String.t(), binary(), {non_neg_integer(), non_neg_integer(), non_neg_integer()}) :: binary()
  def encode_dispatch_compute(shader_source, params, {wg_x, wg_y, wg_z}) do
    src_bytes = :erlang.term_to_binary(shader_source)
    src_len = byte_size(src_bytes)
    params_len = byte_size(params)

    <<0x09, src_len::unsigned-little-32, src_bytes::binary,
      params_len::unsigned-little-32, params::binary,
      wg_x::unsigned-little-32, wg_y::unsigned-little-32, wg_z::unsigned-little-32>>
  end

  @doc "Encode a read_pixels command that reads back GPU texture data."
  @spec encode_read_pixels(non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer()) :: binary()
  def encode_read_pixels(x, y, w, h) do
    <<0x0A, x::unsigned-little-32, y::unsigned-little-32, w::unsigned-little-32, h::unsigned-little-32>>
  end

  @doc "Encode a load_shader command for hot-reloading shaders."
  @spec encode_load_shader(String.t(), String.t()) :: binary()
  def encode_load_shader(name, source) do
    name_bytes = :erlang.term_to_binary(name)
    name_len = byte_size(name_bytes)
    src_bytes = :erlang.term_to_binary(source)
    src_len = byte_size(src_bytes)

    <<0x0B, name_len::unsigned-little-32, name_bytes::binary,
      src_len::unsigned-little-32, src_bytes::binary>>
  end

  @doc "Encode a set_uniform command for shader parameters."
  @spec encode_set_uniform(String.t(), binary()) :: binary()
  def encode_set_uniform(name, data) do
    name_bytes = :erlang.term_to_binary(name)
    name_len = byte_size(name_bytes)
    data_len = byte_size(data)

    <<0x0C, name_len::unsigned-little-32, name_bytes::binary,
      data_len::unsigned-little-32, data::binary>>
  end

  @doc "Encode an image_blit command that draws a loaded image texture."
  @spec encode_image_blit(non_neg_integer(), integer(), integer(), non_neg_integer(), non_neg_integer()) :: binary()
  def encode_image_blit(image_id, x, y, w, h) do
    <<0x16, image_id::unsigned-little-64, x::signed-little-32, y::signed-little-32,
      w::unsigned-little-32, h::unsigned-little-32>>
  end

  @doc "Encode a draw_circle command."
  @spec encode_draw_circle(integer(), integer(), non_neg_integer(), color()) :: binary()
  def encode_draw_circle(cx, cy, radius, color) do
    rgba = color_to_rgba(color)
    <<0x0D, cx::signed-little-32, cy::signed-little-32, radius::unsigned-little-32, rgba::binary>>
  end

  @doc "Encode a fill_circle command."
  @spec encode_fill_circle(integer(), integer(), non_neg_integer(), color()) :: binary()
  def encode_fill_circle(cx, cy, radius, color) do
    rgba = color_to_rgba(color)
    <<0x0E, cx::signed-little-32, cy::signed-little-32, radius::unsigned-little-32, rgba::binary>>
  end

  @doc "Encode a draw_triangle command."
  @spec encode_draw_triangle(integer(), integer(), integer(), integer(), integer(), integer(), color()) :: binary()
  def encode_draw_triangle(x1, y1, x2, y2, x3, y3, color) do
    rgba = color_to_rgba(color)
    <<0x0F, x1::signed-little-32, y1::signed-little-32, x2::signed-little-32,
      y2::signed-little-32, x3::signed-little-32, y3::signed-little-32, rgba::binary>>
  end

  @doc "Encode a fill_triangle command."
  @spec encode_fill_triangle(integer(), integer(), integer(), integer(), integer(), integer(), color()) :: binary()
  def encode_fill_triangle(x1, y1, x2, y2, x3, y3, color) do
    rgba = color_to_rgba(color)
    <<0x10, x1::signed-little-32, y1::signed-little-32, x2::signed-little-32,
      y2::signed-little-32, x3::signed-little-32, y3::signed-little-32, rgba::binary>>
  end

  @doc "Encode a draw_round_rect command."
  @spec encode_draw_round_rect(non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer(), color()) :: binary()
  def encode_draw_round_rect(x, y, w, h, radius, color) do
    rgba = color_to_rgba(color)
    <<0x11, x::unsigned-little-32, y::unsigned-little-32, w::unsigned-little-32,
      h::unsigned-little-32, radius::unsigned-little-32, rgba::binary>>
  end

  @doc "Encode a fill_round_rect command."
  @spec encode_fill_round_rect(non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer(), color()) :: binary()
  def encode_fill_round_rect(x, y, w, h, radius, color) do
    rgba = color_to_rgba(color)
    <<0x12, x::unsigned-little-32, y::unsigned-little-32, w::unsigned-little-32,
      h::unsigned-little-32, radius::unsigned-little-32, rgba::binary>>
  end

  @doc "Encode a set_clip command."
  @spec encode_set_clip(non_neg_integer(), non_neg_integer(), non_neg_integer(), non_neg_integer(), boolean()) :: binary()
  def encode_set_clip(x, y, w, h, enabled) do
    flag = if enabled, do: 1, else: 0
    <<0x13, x::unsigned-little-32, y::unsigned-little-32, w::unsigned-little-32,
      h::unsigned-little-32, flag::unsigned-little-8>>
  end

  @doc "Encode a reset_clip command."
  @spec encode_reset_clip() :: binary()
  def encode_reset_clip do
    <<0x14>>
  end

  @doc "Encode a batch command containing multiple sub-commands."
  @spec encode_batch([binary()]) :: binary()
  def encode_batch(commands) do
    data = :erlang.list_to_binary(commands)
    count = length(commands)
    <<0x15, count::unsigned-little-32, data::binary>>
  end

  # ── Color encoding ────────────────────────────────────────────────────────

  @doc "Convert a color to a 4-byte RGBA binary."
  @spec color_to_rgba(color()) :: binary()
  def color_to_rgba(:black), do: <<0, 0, 0, 255>>
  def color_to_rgba(:white), do: <<255, 255, 255, 255>>
  def color_to_rgba(:red), do: <<255, 0, 0, 255>>
  def color_to_rgba(:green), do: <<0, 255, 0, 255>>
  def color_to_rgba(:blue), do: <<0, 0, 255, 255>>
  def color_to_rgba(:transparent), do: <<0, 0, 0, 0>>
  def color_to_rgba({r, g, b, a}), do: <<r, g, b, a>>
  def color_to_rgba({r, g, b}), do: <<r, g, b, 255>>
end

File: native/dala_gpu/src/nif/command.rs
//! Binary command decoder for the NIF bridge.
//!
//! Decodes the binary command format produced by `Dala.Gpu.Command` into
//! `RenderCommand` values for the render thread.

use crate::RenderCommand;

/// Decode a binary command from the Elixir side into a `RenderCommand`.
///
/// Binary format (matches `Dala.Gpu.Command` encoding):
/// - 0x01: Clear (4 bytes RGBA)
/// - 0x02: FillRect (16 bytes: x,y,w,h as u32 LE + 4 bytes RGBA)
/// - 0x03: DrawLine (16 bytes: x1,y1,x2,y2 as i32 LE + 4 bytes RGBA)
/// - 0x04: Blit (12 bytes: sprite_id as u64 LE + x,y as i32 LE)
/// - 0x05: Present (0 bytes)
/// - 0x06: Resize (8 bytes: width, height as u32 LE)
/// - 0x07: LoadSprite (16 bytes: id as u64 LE + w,h as u32 LE + pixel data)
/// - 0x08: RemoveSprite (8 bytes: id as u64 LE)
pub fn decode_command(data: &[u8]) -> RenderCommand {
    if data.is_empty() {
        return RenderCommand::Present;
    }

    match data[0] {
        0x01 => {
            // Clear
            let color = if data.len() >= 5 {
                [data[1], data[2], data[3], data[4]]
            } else {
                [0, 0, 0, 255]
            };
            RenderCommand::Clear { color }
        }
        0x02 => {
            // FillRect: x,y,w,h as u32 LE + 4 bytes RGBA
            if data.len() >= 21 {
                let x = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let y = u32::from_le_bytes(data[5..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let color = [data[17], data[18], data[19], data[20]];
                RenderCommand::FillRect { x, y, w, h, color }
            } else {
                RenderCommand::Present
            }
        }
        0x03 => {
            // DrawLine: x1,y1,x2,y2 as i32 LE + 4 bytes RGBA
            if data.len() >= 21 {
                let x1 = i32::from_le_bytes(data[1..5].try_into().unwrap());
                let y1 = i32::from_le_bytes(data[5..9].try_into().unwrap());
                let x2 = i32::from_le_bytes(data[9..13].try_into().unwrap());
                let y2 = i32::from_le_bytes(data[13..17].try_into().unwrap());
                let color = [data[17], data[18], data[19], data[20]];
                RenderCommand::DrawLine {
                    x1,
                    y1,
                    x2,
                    y2,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x04 => {
            // Blit: sprite_id as u64 LE + x,y as i32 LE
            if data.len() >= 17 {
                let sprite_id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                let x = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let y = u32::from_le_bytes(data[13..17].try_into().unwrap());
                RenderCommand::Blit { sprite_id, x, y }
            } else {
                RenderCommand::Present
            }
        }
        0x05 => RenderCommand::Present,
        0x06 => {
            // Resize: width, height as u32 LE
            if data.len() >= 9 {
                let width = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let height = u32::from_le_bytes(data[5..9].try_into().unwrap());
                RenderCommand::Resize { width, height }
            } else {
                RenderCommand::Present
            }
        }
        0x07 => {
            // LoadSprite: id as u64 LE + w,h as u32 LE + pixel data
            if data.len() >= 17 {
                let id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let pixel_data = data[17..].to_vec();
                RenderCommand::LoadSprite {
                    id,
                    w,
                    h,
                    data: pixel_data,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x08 => {
            // RemoveSprite: id as u64 LE
            if data.len() >= 9 {
                let id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                RenderCommand::RemoveSprite { id }
            } else {
                RenderCommand::Present
            }
        }
        0x09 => {
            // DispatchCompute: shader_source_len as u32 LE + shader_source + params_len as u32 LE + params + workgroup_count (3xu32 LE)
            if data.len() >= 5 {
                let src_len = u32::from_le_bytes(data[1..5].try_into().unwrap()) as usize;
                if data.len() >= 5 + src_len + 4 {
                    let shader_source = String::from_utf8_lossy(&data[5..5 + src_len]).to_string();
                    let params_offset = 5 + src_len;
                    let params_len = u32::from_le_bytes(
                        data[params_offset..params_offset + 4].try_into().unwrap(),
                    ) as usize;
                    let params = data[params_offset + 4..params_offset + 4 + params_len].to_vec();
                    let wg_offset = params_offset + 4 + params_len;
                    if data.len() >= wg_offset + 12 {
                        let wg_x =
                            u32::from_le_bytes(data[wg_offset..wg_offset + 4].try_into().unwrap());
                        let wg_y = u32::from_le_bytes(
                            data[wg_offset + 4..wg_offset + 8].try_into().unwrap(),
                        );
                        let wg_z = u32::from_le_bytes(
                            data[wg_offset + 8..wg_offset + 12].try_into().unwrap(),
                        );
                        RenderCommand::DispatchCompute {
                            shader_source,
                            params,
                            workgroup_count: (wg_x, wg_y, wg_z),
                        }
                    } else {
                        RenderCommand::Present
                    }
                } else {
                    RenderCommand::Present
                }
            } else {
                RenderCommand::Present
            }
        }
        0x0A => {
            // ReadPixels: x,y,w,h as u32 LE
            if data.len() >= 17 {
                let x = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let y = u32::from_le_bytes(data[5..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                RenderCommand::ReadPixels { x, y, w, h }
            } else {
                RenderCommand::Present
            }
        }
        0x0B => {
            // LoadShader: name_len as u32 LE + name + source_len as u32 LE + source
            if data.len() >= 5 {
                let name_len = u32::from_le_bytes(data[1..5].try_into().unwrap()) as usize;
                if data.len() >= 5 + name_len + 4 {
                    let name = String::from_utf8_lossy(&data[5..5 + name_len]).to_string();
                    let src_offset = 5 + name_len;
                    let src_len =
                        u32::from_le_bytes(data[src_offset..src_offset + 4].try_into().unwrap())
                            as usize;
                    let source =
                        String::from_utf8_lossy(&data[src_offset + 4..src_offset + 4 + src_len])
                            .to_string();
                    RenderCommand::LoadShader { name, source }
                } else {
                    RenderCommand::Present
                }
            } else {
                RenderCommand::Present
            }
        }
        0x0C => {
            // SetUniform: name_len as u32 LE + name + data_len as u32 LE + data
            if data.len() >= 5 {
                let name_len = u32::from_le_bytes(data[1..5].try_into().unwrap()) as usize;
                if data.len() >= 5 + name_len + 4 {
                    let name = String::from_utf8_lossy(&data[5..5 + name_len]).to_string();
                    let data_offset = 5 + name_len;
                    let data_len =
                        u32::from_le_bytes(data[data_offset..data_offset + 4].try_into().unwrap())
                            as usize;
                    let uniform_data = data[data_offset + 4..data_offset + 4 + data_len].to_vec();
                    RenderCommand::SetUniform {
                        name,
                        data: uniform_data,
                    }
                } else {
                    RenderCommand::Present
                }
            } else {
                RenderCommand::Present
            }
        }
        0x0D => {
            // DrawCircle: cx,cy as i32 LE + radius as u32 LE + 4 bytes RGBA
            if data.len() >= 17 {
                let cx = i32::from_le_bytes(data[1..5].try_into().unwrap());
                let cy = i32::from_le_bytes(data[5..9].try_into().unwrap());
                let radius = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let color = [data[13], data[14], data[15], data[16]];
                RenderCommand::DrawCircle {
                    cx,
                    cy,
                    radius,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x0E => {
            // FillCircle: cx,cy as i32 LE + radius as u32 LE + 4 bytes RGBA
            if data.len() >= 17 {
                let cx = i32::from_le_bytes(data[1..5].try_into().unwrap());
                let cy = i32::from_le_bytes(data[5..9].try_into().unwrap());
                let radius = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let color = [data[13], data[14], data[15], data[16]];
                RenderCommand::FillCircle {
                    cx,
                    cy,
                    radius,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x0F => {
            // DrawTriangle: x1,y1,x2,y2,x3,y3 as i32 LE + 4 bytes RGBA
            if data.len() >= 29 {
                let x1 = i32::from_le_bytes(data[1..5].try_into().unwrap());
                let y1 = i32::from_le_bytes(data[5..9].try_into().unwrap());
                let x2 = i32::from_le_bytes(data[9..13].try_into().unwrap());
                let y2 = i32::from_le_bytes(data[13..17].try_into().unwrap());
                let x3 = i32::from_le_bytes(data[17..21].try_into().unwrap());
                let y3 = i32::from_le_bytes(data[21..25].try_into().unwrap());
                let color = [data[25], data[26], data[27], data[28]];
                RenderCommand::DrawTriangle {
                    x1,
                    y1,
                    x2,
                    y2,
                    x3,
                    y3,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x10 => {
            // FillTriangle: x1,y1,x2,y2,x3,y3 as i32 LE + 4 bytes RGBA
            if data.len() >= 29 {
                let x1 = i32::from_le_bytes(data[1..5].try_into().unwrap());
                let y1 = i32::from_le_bytes(data[5..9].try_into().unwrap());
                let x2 = i32::from_le_bytes(data[9..13].try_into().unwrap());
                let y2 = i32::from_le_bytes(data[13..17].try_into().unwrap());
                let x3 = i32::from_le_bytes(data[17..21].try_into().unwrap());
                let y3 = i32::from_le_bytes(data[21..25].try_into().unwrap());
                let color = [data[25], data[26], data[27], data[28]];
                RenderCommand::FillTriangle {
                    x1,
                    y1,
                    x2,
                    y2,
                    x3,
                    y3,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x11 => {
            // DrawRoundRect: x,y,w,h as u32 LE + radius as u32 LE + 4 bytes RGBA
            if data.len() >= 25 {
                let x = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let y = u32::from_le_bytes(data[5..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let radius = u32::from_le_bytes(data[17..21].try_into().unwrap());
                let color = [data[21], data[22], data[23], data[24]];
                RenderCommand::DrawRoundRect {
                    x,
                    y,
                    w,
                    h,
                    radius,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x12 => {
            // FillRoundRect: x,y,w,h as u32 LE + radius as u32 LE + 4 bytes RGBA
            if data.len() >= 25 {
                let x = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let y = u32::from_le_bytes(data[5..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let radius = u32::from_le_bytes(data[17..21].try_into().unwrap());
                let color = [data[21], data[22], data[23], data[24]];
                RenderCommand::FillRoundRect {
                    x,
                    y,
                    w,
                    h,
                    radius,
                    color,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x13 => {
            // SetClip: x,y,w,h as u32 LE + enabled as u8
            if data.len() >= 18 {
                let x = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let y = u32::from_le_bytes(data[5..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let enabled = data[17] != 0;
                RenderCommand::SetClip {
                    x,
                    y,
                    w,
                    h,
                    enabled,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x14 => RenderCommand::ResetClip,
        0x16 => {
            // ImageBlit: image_id as u64 LE + x,y as i32 LE + w,h as u32 LE
            if data.len() >= 25 {
                let image_id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                let x = i32::from_le_bytes(data[9..13].try_into().unwrap());
                let y = i32::from_le_bytes(data[13..17].try_into().unwrap());
                let w = u32::from_le_bytes(data[17..21].try_into().unwrap());
                let h = u32::from_le_bytes(data[21..25].try_into().unwrap());
                RenderCommand::ImageBlit {
                    image_id,
                    x,
                    y,
                    w,
                    h,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x17 => {
            // LoadImage: id as u64 LE + w,h as u32 LE + pixel data
            if data.len() >= 17 {
                let id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                let w = u32::from_le_bytes(data[9..13].try_into().unwrap());
                let h = u32::from_le_bytes(data[13..17].try_into().unwrap());
                let pixel_data = data[17..].to_vec();
                RenderCommand::LoadImage {
                    id,
                    w,
                    h,
                    data: pixel_data,
                }
            } else {
                RenderCommand::Present
            }
        }
        0x18 => {
            // RemoveImage: id as u64 LE
            if data.len() >= 9 {
                let id = u64::from_le_bytes(data[1..9].try_into().unwrap());
                RenderCommand::RemoveImage { id }
            } else {
                RenderCommand::Present
            }
        }
        0x15 => {
            // Batch: count as u32 LE + concatenated command data
            if data.len() >= 5 {
                let count = u32::from_le_bytes(data[1..5].try_into().unwrap());
                let batch_data = data[5..].to_vec();
                RenderCommand::Batch {
                    count,
                    data: batch_data,
                }
            } else {
                RenderCommand::Present
            }
        }
        _ => RenderCommand::Present,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_clear() {
        let data = vec![0x01, 255, 0, 0, 255];
        match decode_command(&data) {
            RenderCommand::Clear { color } => assert_eq!(color, [255, 0, 0, 255]),
            _ => panic!("expected Clear"),
        }
    }

    #[test]
    fn test_decode_fill_rect() {
        let mut data = vec![0x02];
        data.extend_from_slice(&10u32.to_le_bytes());
        data.extend_from_slice(&20u32.to_le_bytes());
        data.extend_from_slice(&100u32.to_le_bytes());
        data.extend_from_slice(&200u32.to_le_bytes());
        data.extend_from_slice(&[0, 255, 0, 255]);

        match decode_command(&data) {
            RenderCommand::FillRect { x, y, w, h, color } => {
                assert_eq!(x, 10);
                assert_eq!(y, 20);
                assert_eq!(w, 100);
                assert_eq!(h, 200);
                assert_eq!(color, [0, 255, 0, 255]);
            }
            _ => panic!("expected FillRect"),
        }
    }

    #[test]
    fn test_decode_present() {
        let data = vec![0x05];
        match decode_command(&data) {
            RenderCommand::Present => {}
            _ => panic!("expected Present"),
        }
    }

    #[test]
    fn test_decode_empty() {
        let data: Vec<u8> = vec![];
        match decode_command(&data) {
            RenderCommand::Present => {}
            _ => panic!("expected Present for empty data"),
        }
    }
}
